### PR TITLE
Add BestValueStagnationEvaluator

### PR DESCRIPTION
**Pull request by smygw72**
## Motivation

Implemented BestValueStagnationEvaluator (referred as Conv-i in ["Automatic Termination for Hyperparameter Optimization"](https://assets.amazon.science/74/fb/591faa624a7c8cd3ccaad8076ee1/automatic-termination-for-hyperparameter-optimization.pdf))
This PR is a part of #4818 and there remain some evaluators (e.g., Expected Improvement, Probability of improvement) to be implemented. 

## Description of the changes

Added BestValueStagnationEvaluator and its tests.

### CODE REVIEW

File: optuna/terminator/improvement/evaluator.py [hunk 1]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
                 "The intersection search space is empty. This condition is not supported by "
                 f"{cls.__name__}."
             )
+
+
+@experimental_class("3.2.0")
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:133]: A nit, but the next release will be `3.4.0`.
```suggestion
@experimental_class("3.4.0")
```

File: optuna/terminator/improvement/evaluator.py [hunk 2]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
...
+
+    Example Usage:
+        >>> evaluator = BestValueStagnationEvaluator(max_stagnation_trials=30)
+        >>> stagnated_trials = evaluator.evaluate()
+    """
+
+    def __init__(
+        self,
+        max_stagnation_trials: int = 30,
+    ):
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:153]: A nit.
```suggestion
    ) -> None:
```

File: optuna/terminator/improvement/evaluator.py [hunk 3]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
...
+        if max_stagnation_trials < 0:
+            raise ValueError("The number of tolerance of step must not be negative.")
+        self._max_stagnation_trials = max_stagnation_trials
+
+    def evaluate(
+        self,
+        trials: List[FrozenTrial],
+        study_direction: StudyDirection,
+    ) -> int:
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:162]: A nit, but we want to follow the return type of the base class.
```suggestion
    ) -> float:
```

File: optuna/terminator/improvement/evaluator.py [hunk 4]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
...
+        best_step = 0
+        for i_step, trial in enumerate(trials):
+            best_value = trials[best_step].value
+            i_step_value = trial.value
+            assert best_value is not None
+            assert i_step_value is not None
+            if is_maximize_direction and (best_value < i_step_value):
+                best_step = i_step
+            elif (not is_maximize_direction) and (best_value > i_step_value):
+                best_step = i_step
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:178]: As `i_step` is not referred to outside of this loop, changing it to a more generic name like `i` would likely make the code more readable.
Similarly, renaming `i_step_value` to `current_value` would make the contrast with `best_value` clearer.

```suggestion
        for i, trial in enumerate(trials):
            best_value = trials[best_step].value
            current_value = trial.value
            assert best_value is not None
            assert current_value is not None
            if is_maximize_direction and (best_value < current_value):
                best_step = i
            elif (not is_maximize_direction) and (best_value > i_step_value):
                best_step = i
```

File: optuna/terminator/improvement/evaluator.py [hunk 5]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
...
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed toevaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
+
+    Args:
+        - max_stagnation_trials (int): The maximum number of trials allowed for stagnation.
+
+    Example Usage:
+        >>> evaluator = BestValueStagnationEvaluator(max_stagnation_trials=30)
+        >>> stagnated_trials = evaluator.evaluate()
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:147]: Could you follow the Optuna's  docstring style, please?

https://github.com/optuna/optuna/blob/9f0be552b618c695afd5569c81fb45140c1613f2/optuna/terminator/improvement/evaluator.py#L52-L62

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your great PR. Let me share my early comments.

File: optuna/terminator/improvement/evaluator.py [hunk 6]
```diff
@@ -128,3 +128,53 @@ def _validate_input(
...
+
+
+@experimental_class("3.4.0")
+class BestValueStagnationEvaluator(BaseImprovementEvaluator):
+    """
+    Evaluates the stagnation period of the best value in an optimization process.
+
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed toevaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:140]: @g-votte What do you think of this class name?
If you have time, could you join the review, please?

File: optuna/terminator/improvement/evaluator.py (hunk 5 above)
Comment (smygw72) [reply optuna/terminator/improvement/evaluator.py]: Updated as follow.
```
    """Evaluates the stagnation period of the best value in an optimization process.

    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
    and is designed toevaluate the remaining trials before reaching this maximum period
    of allowed stagnation.

    Args:
        max_stagnation_trials (int):
            The maximum number of trials allowed for stagnation.
    """
```

File: tests/terminator_tests/improvement_tests/test_evaluator.py [hunk 7]
```diff
@@ -60,11 +61,25 @@ def test_evaluate() -> None:
...
+
+
 def test_evaluate_with_no_trial() -> None:
-    evaluator = RegretBoundEvaluator()
+    evaluators = [RegretBoundEvaluator(), BestValueStagnationEvaluator()]
 
     with pytest.raises(ValueError):
-        evaluator.evaluate(trials=[], study_direction=StudyDirection.MAXIMIZE)
+        for evaluator in evaluators:
+            evaluator.evaluate(trials=[], study_direction=StudyDirection.MAXIMIZE)
```
Comment (toshihikoyanase) [inline tests/terminator_tests/improvement_tests/test_evaluator.py:82]: Following the convention of this codebase, we want to use `@pytest.mark.parametrize`, for example, as follows:


```python
@pytest.mark.parametrize(
    "evaluator", [RegretBoundEvaluator(), BestValueStagnationEvaluator()])
)
def test_evaluate_with_no_trial(evaluator: BaseImprovementEvaluator) -> None:

    with pytest.raises(ValueError):
        evaluator.evaluate(trials=[], study_direction=StudyDirection.MAXIMIZE)
```

File: tests/terminator_tests/improvement_tests/test_evaluator.py [hunk 8]
```diff
@@ -60,11 +61,25 @@ def test_evaluate() -> None:
...
+    evaluator = BestValueStagnationEvaluator(max_stagnation_trials=1)
+    trials = [create_trial(value=value) for value in [0, 1, 2]]
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MAXIMIZE) == 1
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MINIMIZE) == -1
+    trials = [create_trial(value=value) for value in [0, 1, 0]]
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MAXIMIZE) == 0
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MINIMIZE) == -1
+    trials = [create_trial(value=value) for value in [0, 0, 0]]
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MAXIMIZE) == -1
+    assert evaluator.evaluate(trials=trials, study_direction=StudyDirection.MINIMIZE) == -1
```
Comment (toshihikoyanase) [inline tests/terminator_tests/improvement_tests/test_evaluator.py:74]: `evaluator.evaluate` always return -1 if the study direction is minimize. Could you verify it returns positive values, please?

**Review by toshihikoyanase — COMMENTED** [review]
I added some comments on the unit tests.

File: optuna/terminator/improvement/evaluator.py [hunk 9]
```diff
@@ -128,3 +128,49 @@ def _validate_input(
...
+@experimental_class("3.4.0")
+class BestValueStagnationEvaluator(BaseImprovementEvaluator):
+    """Evaluates the stagnation period of the best value in an optimization process.
+
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed toevaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
+
+    Args:
+        max_stagnation_trials (int):
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:142]: Thank you for your update. The type information will be added by sphinx, so could you remove it, please?
```suggestion
        max_stagnation_trials:
```

File: optuna/terminator/improvement/evaluator.py [hunk 10]
```diff
@@ -128,3 +128,49 @@ def _validate_input(
...
+        max_stagnation_trials:
+            The maximum number of trials allowed for stagnation.
+    """
+
+    def __init__(
+        self,
+        max_stagnation_trials: int = 30,
+    ) -> None:
+        if max_stagnation_trials < 0:
+            raise ValueError("The number of tolerance of step must not be negative.")
```
Comment (toshihikoyanase) [inline optuna/terminator/improvement/evaluator.py:151]: `The number of tolerance of step` might not remind `max_stagnation_trials`. Could you update the message, for example, as follows:

```suggestion
            raise ValueError("The maximum number of stagnant trials must not be negative.")
```

File: optuna/terminator/__init__.py [hunk 11]
```diff
@@ -16,6 +17,7 @@
     "report_cross_validation_scores",
     "StaticErrorEvaluator",
     "BaseImprovementEvaluator",
+    "BestValueStagnationEvaluator",
```
Comment (toshihikoyanase) [inline optuna/terminator/__init__.py:20]: Could you add `BastValueStagnationEvaluator` to the reference, please?

https://github.com/optuna/optuna/blob/master/docs/source/reference/terminator.rst

File: tests/terminator_tests/improvement_tests/test_evaluator.py (hunk 8 above)
Comment (smygw72) [reply tests/terminator_tests/improvement_tests/test_evaluator.py:76]: Added test cases
https://github.com/optuna/optuna/pull/4974/files#diff-09080e058d6651a7a3855dfbf9de39ff4464796da0b2ad025e8e9adac546e7d1R64-R76

File: tests/terminator_tests/improvement_tests/test_evaluator.py (hunk 7 above)
Comment (smygw72) [reply tests/terminator_tests/improvement_tests/test_evaluator.py]: Done

File: optuna/terminator/__init__.py (hunk 11 above)
Comment (smygw72) [reply optuna/terminator/__init__.py:20]: Done

**Comment by not522** [comment]
@contramundum53 Could you review this PR?

File: optuna/terminator/improvement/evaluator.py [hunk 12]
```diff
@@ -128,3 +128,58 @@ def _validate_input(
...
             )
+
+
+@experimental_class("3.4.0")
+class BestValueStagnationEvaluator(BaseImprovementEvaluator):
+    """Evaluates the stagnation period of the best value in an optimization process.
+
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed to evaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
```
Comment (contramundum53) [inline optuna/terminator/improvement/evaluator.py:139]: I think it should be clearly documented that this evaluator is to be used with `StaticErrorEvaluator(0)`.

Comment (contramundum53) [reply optuna/terminator/improvement/evaluator.py:139]: Maybe it would be kind to add an example?

Comment (toshihikoyanase) [reply optuna/terminator/improvement/evaluator.py:139]: @contramundum53 That sounds reasonable. However, those appear to be optional items, so we can address them in follow-up PRs.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. It looks excellent to me. I'm really grateful you've implemented the new Terminator algorithm.

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Pull request merged**